frame/evm/src/lib.rs

```diff
@@ -58,41 +57,121 @@ impl FeeCalculator for () {
-/// `H160` for EVM module.
-///
-/// Accounts and contracts of this module are stored in its own
-/// storage, in an Ethereum-compatible format. In order to communicate
-/// with the rest of Substrate module, we require an one-to-one
-/// mapping of Substrate account to Ethereum address.
-pub trait ConvertAccountId<A> {
-	/// Given a Substrate address, return the corresponding Ethereum address.
-	fn convert_account_id(account_id: &A) -> H160;
+pub trait EnsureAddressOrigin<OuterOrigin> {
+	/// Success return type.
+	type Success;
+
+	/// Perform the origin check.
+	fn ensure_address_origin(
+		address: &H160,
+		origin: OuterOrigin,
+	) -> Result<Self::Success, BadOrigin> {
+		Self::try_address_origin(address, origin).map_err(|_| BadOrigin)
+	}
+
+	/// Try with origin.
+	fn try_address_origin(
+		address: &H160,
+		origin: OuterOrigin,
+	) -> Result<Self::Success, OuterOrigin>;
 }
 
-/// Hash and then truncate the account id, taking the last 160-bit as the Ethereum address.
-pub struct HashTruncateConvertAccountId<H>(PhantomData<H>);
+/// Ensure that the EVM address is the same as the Substrate address. This only works if the account
+/// ID is `H160`.
+pub struct EnsureAddressSame;
+
+impl<OuterOrigin> EnsureAddressOrigin<OuterOrigin> for EnsureAddressSame where
+	OuterOrigin: Into<Result<RawOrigin<H160>, OuterOrigin>> + From<RawOrigin<H160>>,
+{
+	type Success = H160;
+
+	fn try_address_origin(
+		address: &H160,
+		origin: OuterOrigin,
+	) -> Result<H160, OuterOrigin> {
+		origin.into().and_then(|o| match o {
+			RawOrigin::Signed(who) if &who == address => Ok(who),
+			r => Err(OuterOrigin::from(r))
+		})
+	}
+}
 
-impl<H: Hasher> Default for HashTruncateConvertAccountId<H> {
-	fn default() -> Self {
-		Self(PhantomData)
+/// Ensure that the origin is root.
+pub struct EnsureAddressRoot<AccountId>(sp_std::marker::PhantomData<AccountId>);
+
+impl<OuterOrigin, AccountId> EnsureAddressOrigin<OuterOrigin> for EnsureAddressRoot<AccountId> where
+	OuterOrigin: Into<Result<RawOrigin<AccountId>, OuterOrigin>> + From<RawOrigin<AccountId>>,
+{
+	type Success = ();
+
+	fn try_address_origin(
+		_address: &H160,
+		origin: OuterOrigin,
+	) -> Result<(), OuterOrigin> {
+		origin.into().and_then(|o| match o {
+			RawOrigin::Root => Ok(()),
+			r => Err(OuterOrigin::from(r)),
+		})
 	}
 }
 
-impl<H: Hasher, A: AsRef<[u8]>> ConvertAccountId<A> for HashTruncateConvertAccountId<H> {
-	fn convert_account_id(account_id: &A) -> H160 {
-		let account_id = H::hash(account_id.as_ref());
-		let account_id_len = account_id.as_ref().len();
-		let mut value = [0u8; 20];
-		let value_len = value.len();
-
-		if value_len > account_id_len {
-			value[(value_len - account_id_len)..].copy_from_slice(account_id.as_ref());
-		} else {
-			value.copy_from_slice(&account_id.as_ref()[(account_id_len - value_len)..]);
-		}
+/// Ensure that the origin never happens.
+pub struct EnsureAddressNever<AccountId>(sp_std::marker::PhantomData<AccountId>);
```

Comment:
Out of curiosity, in which case this is necessary ?

Reply:
This is for "disabling" `CallOrigin` or `WithdrawOrigin`. For example, in identity `H160` account ID case we do not need `withdraw` at all, and so it can set `WithdrawOrigin` to `EnsureAddressNever` and `CallOrigin` to `EnsureAddressSame`.